### ISSUE TITLE
VS Enterprise 2019 stops responding after insta

### ISSUE THREAD

**Issue by MatrixArchitect** [microsoft/cascadia-code#319]
# Environment

```none
Cascadia Code version number: 2007.01
Application (with version) used to display text: VS Enterprise 2019 16.6.3
OS platform and version: Windows Server 2016 1607 Build 14393.3750 running in a VM
Screen resolution (i.e. 220dpi): Windows 10 Host @ 3840x2160 (VM can't change resolution)

Any other software?
A number of plug-ins. Plugs with their own spot on the Extensions menu:
PostSharp
SuperCharger
Let me know if you want a full list of all plug ins

This is a bare bones dev VM. No other tools other than LinqPad, Chrome, IE and Office 2016

# Steps to reproduce

Installed Cascadia TTF fonts for the first time. Changed Editor fonts to Cascadia in VS 2019 Enterprise. Ligatures won't work. So restarted. And then VS would become unresponsive as soon as first code file was loaded in an editor in a solution. VS would be responsive until a solution was opened and a text file was loaded into the editor. I noticed the not equal ligature work in the editor though it was not responsive. Multiple restarts didn't resolve the issue. Finally, restarting VS without opening a solution, and changing the font back to default Consolas font resolved the issue.

# Expected behavior

Visual Studio should not be unresponsive and continue to operate as it did with default Consolas font.

# Actual behavior

Visual Studio become unresponsive. No amount of restarts fixes the issue.

Let me know if you need any further information.

**Comment by aaronbell** [comment]
Strange. I have installed the Cascadia fonts in Visual Studio Enterprise as well and had no such issues.

Are you able to locate error reports from the Event Viewer or elsewise that can explain what might be causing the issue. Also is the font enumerating properly in the font menu (can you see all the weights)?

@DHowett do you know if Windows Server has  issues with variable fonts?

**Comment by DHowett** [comment]
Not to my knowledge! @matrixarchitect can you share a dump (generated in task manager) of devenv.exe? My e-mail address is in my github profile.

You might need to explicitly use the 32-bit task manager to get a 32-bit dump because devenv is a 32-bit process :smile:

**Comment by MatrixArchitect** [comment]
Thanks for responding @aaronbell.

I only see font sizes up to 24 points. See attachment.
<img width="345" alt="FontSize" src="https://user-images.githubusercontent.com/55520144/86418130-02637980-bc84-11ea-9724-5455ee07aa38.PNG">

I have also attached what I see in the Event Viewer. It's referring me to Security and Maintenance in control panel. Not clear though exactly what it wants me to look at.
<img width="948" alt="EventViewer" src="https://user-images.githubusercontent.com/55520144/86418292-8c134700-bc84-11ea-9e20-7ae30c0de058.PNG">
<img width="462" alt="Sys Sec2" src="https://user-images.githubusercontent.com/55520144/86418308-99303600-bc84-11ea-80a2-c589d8957a41.PNG">
<img width="533" alt="Sys Sec" src="https://user-images.githubusercontent.com/55520144/86418315-9df4ea00-bc84-11ea-9126-a2a9b6361466.PNG">

Let me know if there's any other information I can collect for you.

**Comment by MatrixArchitect** [comment]
Thanks @DHowett 

The dump file's too big. Even compressed it's like 337 MB.

Here's a one drive link to it

https://1drv.ms/u/s!Ai6zAEN6QgIqgXnBfSW0slTJrvt5?e=v3O2mj

I also sent you a link via One Drive.

**Comment by DHowett** [comment]
Thanks! I can't promise any action on this bug over the long holiday weekend, but I'll see that the right folks look at this.

**Comment by MatrixArchitect** [comment]
No worries. Take your time. Happy 4th. Enjoy the long weekend.

**Comment by aaronbell** [comment]
Sorry! I don't think I explained what I meant—are you seeing all the different weights of Cascadia Code? As in Cascadia Code ExtraLight, SemiLight, Light, SemiBold, Bold? Or just the one "Cascadia Code"?

**Comment by MatrixArchitect** [comment]
I only see Cascadia Code. I installed all 4 ttf fonts.

<img width="345" alt="FontSize" src="https://user-images.githubusercontent.com/55520144/86421195-eebd1080-bc8d-11ea-8c91-34a30df376c3.PNG">

I installed the font by right clicking on it in windows explorer and selecting install from the context menu (instead of double clicking the font and clicking install in the window that pops up). I hope that doesn't make a difference.

**Comment by MatrixArchitect** [comment]
Ok I uninstalled and reinstalled the fonts by double clicking the font and then clicking install in the window that opens up. Doesn't look like it helped. I also looked at it in %windir%\fonts. It appears that fonts for which you see different weights have a multi-folder icons. Cascadia has a single folder icon.
<img width="376" alt="FontFamily" src="https://user-images.githubusercontent.com/55520144/86421738-f5e51e00-bc8f-11ea-90bd-5ce538ad0f43.PNG">

**Comment by aaronbell** [comment]
Hm. I haven’t looked at the dump you provided (not that I’d be able to tell anything from it anyway) but looking at the images you provided it appears that your system does not support variable fonts. 

I’d expect that the font should still work, though, and shouldn’t be causing crashes. You’d just only have access to the Regular weight (and it appears to be to case, per the fonts folder image).

Alternatively, there could be a font cache problem. Is this the first time you’ve installed Cascadia Code on this VM? Or did you have a version installed previously? Windows doesn’t handle uninstalling and reinstalling fonts as well as I’d like, so that could be a source of problems if there’s a cache issue and Windows is confused between the old and new version of the font.

**Comment by MatrixArchitect** [comment]
This was the first time I installed them. I never had these fonts before.

I followed the instruction on the home page of the fonts and only installed the **true type** fonts TTF.

<img width="679" alt="Installation" src="https://user-images.githubusercontent.com/55520144/86431469-ba574d80-bca9-11ea-8c31-66752ebd9952.PNG">

You mentioned variable fonts. I'm not an expert on fonts, so I did a quick search and it appears that variable fonts are **open type** fonts OTF. Should I have installed the OTF fonts as well? Should I have installed all of them? (TTF, OTF and WOFF2?)

**Comment by aaronbell** [comment]
You followed the instructions correctly—variable fonts can be both OTF and TTF (modernly both are the same backend with slightly different 'flavors'). Basically a variable font is a single font file that contains multiple styles, such as weight, width, etc. For the purposes of Windows devices we provide a TTF variable font. 

Basically, it sounds like your system just doesn't understand what a variable font is, and that may be somehow related to VS crashing. What's confusing is that in my experience, systems that can't deal with variable fonts just see them as Regular weights, so I'm not sure why VS is crashing, but I'm sure the dump will provide more info on that. 

Anyway, it seems to me that we should definitely provide static TTF instances of Cascadia Code, eh @DHowett? Need to figure out how to set up ttfautohint.

**Comment by MatrixArchitect** [comment]
Thanks for the explanation. You may be on to something. There are OS differences. When I double-click on the TTF file on the host Windows 10 machine I see a Next and Previous button in the font window that cycle through the font faces.

![Windows10FontWindow](https://user-images.githubusercontent.com/55520144/86490126-32aa2700-bd1b-11ea-8c1d-79b4e8650454.jpg)


I don't see these Next and Previous buttons on Windows Server 2016 VM. Neither does it mention the font face as in Windows 10.
<img width="599" alt="Windows2016FontWindow" src="https://user-images.githubusercontent.com/55520144/86490111-23c37480-bd1b-11ea-9093-0e2e168a0869.PNG">


In Windows 2016, the fonts were installed in %windir%\fonts folder. In Windows 10, they're installed somewhere else. I don't see them in %windir%\fonts folder. But I do see the font installed in Settings>Personalization>Fonts with all 6 font faces.

**Comment by aaronbell** [comment]
Yeah, looks like Windows Server 2016 does not have support for the format. Sorry for the trouble! I'd still expect the fonts to work, at least at a minimal level, but hopefully when folks are back next week they can figure out what's going on with the dump. 

For now, I'd suggest installing the [previous release of Cascadia Code](https://github.com/microsoft/cascadia-code/releases/tag/v2005.15) which is not a variable font, and should work without any trouble. You won't have access to the full range of weights, but you'll get the ligatures!

**Comment by MatrixArchitect** [comment]
Cool thanks @aaronbell. V2005.15 works without a hiccup. My 2¢: 2007.01 needs to be made safe enough so it falls back gracefully to 2005.15 behavior on systems that don't support variable fonts.

**Comment by MatrixArchitect** [comment]
Ok I spoke too soon. With 2007.01 VS would become unresponsive as soon as I would change the font to Cascadia. That didn't happen with 2005.15. The ligatures worked after I changed the font to Cascadia. I got excited. :D But now VS becomes unresponsive on restart and project load. Have a great weekend and we'll pick up this thread next week.

**Comment by aaronbell** [comment]
A bit of background for you:
Variable fonts are basically like a standard font file, but have additional little bits added in that say, "when you move along axis ZZZ, change the position of point 15 by (x,y)". If a computer doesn't understand the variable font bits, it should just read it as a standard font file with no variation. That's why you didn't see the "Next>" button when you installed it. 

However, there is something else amiss. If the 2005.15 version of the font (which has been out in the wild for over a month and a half with no similar reported issues) is also causing your VS install to be unresponsive, I start to suspect that it might be something on your end 😦. Hopefully the dump files can shed some light on this.

**Comment by MatrixArchitect** [comment]
Re: something wrong on my environment. It's quite possible. However, I am usually up to date on patches, both OS and VS.

I am old fashioned and I still develop on a server in a VM to keep my host OS clean of any development tools. I just don't want to subject the host OS to a lot of a installations and uninstallations which is fairly common in dev environments. If something goes wrong, I just roll back to the previous version of VHDX or worst case, trash the VHDX and restore from backup. I can be up and running again in case of a DR in less than 30 mins. I don't have that luxury if I put all my tools on the host VS which would take a lot longer to rebuild. And on top of that, I have the flexibility of installing full server versions of IIS or SQL etc. should I need to, to say the least.

These days I rarely find developers using a server OS. Most people use Windows 10 and it's good enough for them.  That could be another reason why no one else has whined. But I'll wait to see if the dump analysis reveals anything on my end. Would be good to know if it's on my end, because fixing it is would be in my control. :)

**Comment by DHowett** [comment]
So, I went ahead and jumped in with both feet ...

![image](https://user-images.githubusercontent.com/189190/86626021-f248e400-bf7a-11ea-8e75-cf91da65bce4.png)

**Comment by DHowett** [comment]
Ha! Okay, it's triggered by displaying multiple styles in the buffer. I set comments to be Bold and it locked up the moment I finished introducing a comment.

The hang is somewhere in DirectWrite as it tries to `AccumulateInterpolatedTupleDeltas`.

**Comment by aaronbell** [comment]
@Dhowett, thanks for the investigation! From your description, it sounds like you were testing the Variable font? I was surprised that @MatrixArchitect also said that the older, non-variable version caused hangs as well. Or would this same hang spot be valid for that too?

**Comment by DHowett** [comment]
I believe the same code is involved in fake-bolding the font.

**Comment by aaronbell** [comment]
Ahh, OK. So it is an issue with the fake bold. That'd make sense. So if we provided a set of static TTFs, that should resolve things.

**Comment by DHowett** [comment]
I actually can't get this to reproduce w/ 2005.15. Hm.

**Comment by MatrixArchitect** [comment]
@DHowett for me the hang w/2005.15 happened on VS restart not on changing fonts. I was able to change fonts successfully. Was very happy. Wanted to see what happens when I restart VS. And I was happy no more. :)

**Comment by DHowett** [comment]
That would be very helpful, as I wasn't able to break it even on relaunch.
Thanks :smile:

**Comment by MatrixArchitect** [comment]
@DHowett Thank you again. Here's the link to dump for 2005.15. I've also sent you an email via OneDrive. 
https://1drv.ms/u/s!Ai6zAEN6QgIqgXsvptJMkDq77C9p?e=0juIWi

I can't claim I understood everything you guys said earlier about mixing fonts styles and their interactions. But I do wonder if it is worth pointing out that I am using a (mini ReSharper like) plugin called SuperCharger that selectively mucks around with font styles on keywords and code components etc. Could that be an issue? See screenshot.

<img width="503" alt="SuperchargerFonts" src="https://user-images.githubusercontent.com/55520144/86849123-0a486280-c064-11ea-924f-f616a54a2891.PNG">

**Comment by DHowett** [comment]
Interesting. That could be related.

Unfortunately, there was nothing interesting in the most recent dump you shared. Was VS stalled when you took it? It didn't seem like any of its 34 threads were doing anything particularly interesting. ☹️

**Comment by MatrixArchitect** [comment]
Yes it was stalled. It was stalled indeed.

If you like, I can try to reproduce it and take another dump and a screenshot saying "(Not Responding)" in the title bar. Would that help?

**Comment by DHowett** [comment]
No, no, I just had to make sure :smile: thanks for all the help so far.
I’ve looped one a couple folks from the font rendering team on my side, because there shouldn’t be a way for a font to take down a process. I’m hoping the response is “here’s what’s wrong with your font” rather than “that’s a bug in Windows”!

**Comment by MatrixArchitect** [comment]
I also hope it's not a windows bug.

The way I see it, I am the one getting help. So thank you for that. I am just assisting in data collection to make this a better product for the entire dev community out there. Let me know if there's anything else I can do to assist with the forensics.

**Comment by aaronbell** [comment]
@DHowett any word back?

**Comment by Hoaas** [comment]
Wow.
Also experienced this bug. Took a few days of troubleshooting and Googling before I found this. After looking at ProcMon, I saw one of the last things VS did before freezing was reading the font files.

Guess it's time to upgrade the OS ...

- Cascadia Code 2019.4, CascadiaCodePL.ttf
- Visual Studio 2019 16.7.4
- Windows Server 2016, Version 1607 (OS Build 14393.3930)

**Comment by ITSecMedia** [comment]
I struggled 2 days with the bug in the latest release **Cascadia Code 2009.22** (static TTF) until I tried a previous release. 
The current release (Cascadia Code 2009.22) is no good if you want to use the static TTF build with Intellij IDEA as the font variations doesn't get detected anymore. This one works -> CascadiaCode-2009.14

**Comment by aaronbell** [comment]
@DHowett Ping :) Any word on this one?

**Comment by aaronbell** [comment]
@DHowett Ping :)
